Restore payment amount fuzzing

darosior opened the pull request:
[It got stuck into the red part of a diff](https://github.com/ElementsProject/lightning/pull/2215/files#diff-fc86e59a6aa3bf284fbaf6be349102d7L536)

niftynei commented on plugins/pay.c:
```diff
@@ -1061,6 +1067,11 @@ static struct command_result *json_pay(struct command *cmd,
 		pc->msat = *msat;
 	}
 
+	if (*fuzzpercent > 0.0) {
```
though there's a comment on the old version of fuzzing that mentions that the total amount doesn't exceed the allowed limit for a payment. i wonder if this is still implemented also

darosior replied:
Hmm yeah I wondered to make it a `do {} while (overpayment > maxfeepercent * amount)`. But the default `fuzzpercent <= maxpercent` and since `fuzzpercent` and `maxpercent` are optional, I think someone providing them knows what they doing ?

rustyrussell commented:
I wonder if we should, instead, use the shadow route to fuzz?  We use that to fuzz the cltv, should we use that to fuzz the amount?

I'd like to generally make our shadow route stronger, and this would be a good start...

darosior commented:
> I wonder if we should, instead, use the shadow route to fuzz? We use that to fuzz the cltv, should we use that to fuzz the amount?

Thinking about it, I said myself it would be harder to test : this means it might be harder to analyze as well, I'll give it a try!

darosior commented:
Did it through shadow routing, added a FIXME in `pay` because the `cltv` delta in `route_info` is an `u16` (why?) and I didn't want to go down the rabbit hole to change it in this PR. Added an other one in the shadow routing test as testing the cltv delta needs the `pay` command to return it and it seems outside the scope of this PR.

rustyrussell commented on CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- JSON API: The `pay` command now accepts a new parameter, `fuzzpercent` used to fuzz the fees for route calculation.
```
I've been looking at this metric, and it doesn't really make sense.  Fees are so tiny, that 5% is hardly a change at all, and seems to contribute almost nothing to route diversity :(   But we don't want to increase it so much that we breach 'maxfeepercent' which people really care about!

I'd be happier with moving the routing algo towards randomizing the route so that it spends up to min(maxfeepercent / 2, best-fee * 2).  i.e. if we're well under our `maxfeepercent`, then you can spend twice as much as the best case on route obfuscation.

Claudia Diaz made some pointed remarks about our route selection in Lightning, and I'd really like to improve it.

Anyway, this is an aside: I'd prefer to avoid exposing this for now.

darosior replied:
Ok, will remove this commit then.

darosior replied:
> I'd be happier with moving the routing algo towards randomizing the route so that it spends up to min(maxfeepercent / 2, best-fee * 2).

Is this part of the reason why you proposed higher fees, to have more chances to fuzz route computation ?

rustyrussell commented on plugins/pay.c:
```diff
@@ -824,25 +824,32 @@ static struct command_result *add_shadow_route(struct command *cmd,
 	size_t i;
 	u64 sample = 0;
-	u32 cltv, best_cltv;
+	/* FIXME: Use route_info's cltv_expiry_delta member instead */
```
The gossip protocol restricts that to a u16 btw...

darosior replied:
Didn't notice that subtlety, thanks

rustyrussell commented on plugins/pay.c:
```diff
@@ -103,6 +103,9 @@ struct pay_command {
 
+	/* Disable the use of shadow route ? */
+	double use_shadow;
+
```
Can we make this dev-only please?  I don't want normal people to disable shadow routing, it's a terrible idea.

darosior replied:
Will do

darosior commented:
- Dropped the first commit (`fuzzpercent` parameter).
- Added a json helper (basically a copy of the above one, `json_to_number`) for `u16` and used it for cltv (that fixed the previously introduced FIXME).
- Made `use_shadow` dev-only, which introduced a some changes since it's a parameter of a non-dev command. Users are now shadow-safe !
(and rebased on master)

Thanks for the review !

darosior commented:
Ah I forgot the CHANGELOG trick, will do now
